hashintel/hash
# H-1860, H-1859, H-1858, H-1861, H-2196: Ability to set feature flags/entitlements per user + redirect from user entity page

## 🌟 What is the purpose of this PR?




This PR adds a `enabledFeatureFlag` property to the `User` entity, which is an array of strings that identify enabled feature flags. Initially the supported feature flags are:
- `Pages` - whether or not pages can be viewed in the sidebar or as a pinned tab in a user profile
- `Documents` - whether or not Document pages can be created or edited
- `Canvases` - whether or not Canvas pages can be created or edited
- `Notes` - whether the notes page can be accessed, viewed in the sidebar, or whether notes can be created in the browser plugin

⚠️ When this PR merges, all users will have the `enabledFeatureFlags` property as `undefined` in production. This means the above feature flags will be disabled, i.e. pages, documents, canvas, notes functionality becomes inaccessible to everyone.

As a drive-by:

- redirects users from the entity editor to the profile page when viewing a `User` entity to discourage users from directly modifying user entity properties (H-2196)

## 🔗 Related links





- H-1860
- H-1859
- H-1858
- H-1861
- H-2196

## 🔍 What does this change?




- In the developer environment, Alice is seeded with all feature flags enabled, whereas Bob is seeded with no feature flags enabled

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



## 🐾 Next steps



- provide a `/admin` UI for instance admins to set which feature flags are enabled/disabled for users (H-2194)

## 🛡 What tests cover this?



None

## ❓ How to test this?



1. Checkout the branch and start the local developer environment
2. Login as `Alice`, and ensure Pages/Canvases/Documents/Notes are accessible as they were previously
3. Login as `Bob`, and observe the relevant functionality disappearing for each feature flag
4. Run the browser extension, and ensure the quick notes section is hidden for `Bob` but available for `Alice`
5. Login as the instance admin and ensure the feature flags are all enabled by default (despite the `enabledFeatureFlag` property not being set on that user's entity)

## 📹 Demo
